Mathlib/Tactic/Linter/Lint.lean

```diff
@@ -269,4 +269,70 @@ initialize addLinter longLineLinter
 
 end LongLine
 
+/-! # The "openClassical" linter -/
+
+/-- The "openClassical" linter emits a warning on `open Classical` statements which are not
+scoped to a single declaration. This can hide that some theorem statements would be better stated
+with explicit decidability statements.
+-/
+register_option linter.openClassical : Bool := {
+  defValue := true
+  descr := "enable the openClassical linter"
+}
+
+namespace OpenClassical
+
+/-- Gets the value of the `linter.openClassical` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.openClassical o
+
+/-- If `stx` is syntax describing an `open` command, `extractOpenNames stx`
+returns syntax for the opened names as well as an array of all names opened
+(discarding renamed or hidden items). -/
+def extractOpenNames : Syntax → Option (Syntax × (Array Name)) := fun stx ↦ do
+  match stx with
+    | `(command|open $openHiding:openHiding) =>
+      -- The first argument is the identifier we care about.
+      let arg := (openHiding.raw.getArgs).get! 0
+      some (arg, #[arg.getId])
+    | `(command|open $openRenaming:openRenaming) =>
+      -- The first argument is the identifier we care about.
+      let arg := (openRenaming.raw.getArgs).get! 0
+      some (arg, #[arg.getId])
+    | `(command|open $openOnly:openOnly) =>
+      -- The first argument is the identifier we care about.
+      some (openOnly.raw, #[(openOnly.raw.getArgs.get! 0).getId])
+    | `(command|open $openDecl:openSimple) =>
+      -- The first and only argument of `openDecl` is an array containing all the names we
+      -- are interested in.
+      let namesSyntax := (openDecl.raw.getArgs).get! 0
+      some (namesSyntax, (namesSyntax.getArgs).map fun a ↦ a.getId)
+    | `(command|open $openScoped:openScoped) =>
+      -- The first argument of `openScoped` is "scoped",
+      -- the second one is an array containing all the names we are interested in.
+      let namesSyntax := (openScoped.raw.getArgs).get! 1
+      some (namesSyntax, (namesSyntax.getArgs).map fun a ↦ a.getId)
+    | _ => none
+
+@[inherit_doc Mathlib.Linter.linter.openClassical]
+def openClassicalLinter : Linter where run := withSetOptionIn fun stx ↦ do
+    unless getLinterHash (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+     return
+    -- TODO: once mathlib's Lean version includes leanprover/lean4#4741, make this configurable
+    unless #[`Mathlib, `test, `Archive, `Counterexamples].contains (← getMainModule).getRoot do
+      return
+    -- If `stx` describes an `open` command, extract the list of opened namespaces.
+    if let some (stxN, names) := extractOpenNames stx then
+      if names.contains `Classical then
+        Linter.logLint linter.openClassical stxN "\
```

Comment:
If in `extractOpenNames` you return an `Array Syntax` as second factor, then you can log the warning just at the `Classical`, rather than the full list of `open`ed namespaces (assuming I am parsing the code correctly).

Reply:
You are, this is possible. I considered it - since you also asked, I just changed the code accordingly. The diff is neutral, so this is probably a good change.